src/btree/bt_compact.c

```diff
@@ -173,23 +173,24 @@ __compact_page(WT_SESSION_IMPL *session, WT_REF *ref, bool *skipp)
     WT_REF_LOCK(session, ref, &previous_state);
 
     /*
-     * Skip deleted pages but consider them progress (the on-disk block is discarded by the next
-     * checkpoint).
-     */
-    if (previous_state == WT_REF_DELETED)
-        *skipp = false;
+     * Skip obsolete deleted pages (the on-disk blocks are already discarded) */
+    if (previous_state == WT_REF_DELETED && ref->page_del == NULL)
+        *skipp = true;
```

Comment:
I don't understand why the logic is changing here. A globally visible deleted page should be reclaimed by the next checkpoint, and that reclamation should make blocks available for subsequent compact passes that could allow for progress to be made.

I went digging, and it appears as though the check was added in WT-6001 (https://github.com/wiredtiger/wiredtiger/pull/7037/files), but I can't see any justification. Actually, looks like this is older than that via WT-2923
https://github.com/wiredtiger/wiredtiger/pull/3053/files
There was a more useful comment that got trimmed down in WT-6001, it mentions "in a better world we'd write the parent to delete the page".
So, mabye that answers my question somewhat? i.e: Would we need to mark the internal page for re-writing in order to actually clean up this deleted page? Is that more possible with the current version of compact than it was in 2016?

Reply:
Yes, that's correct. We may not rewrite the internal page to update the ref to the deleted page. Not skipping here results in us maxing out the number of passes compact does on a table which is 100. This results in 200 checkpoints where we think we have something to reclaim but we just continue to see the same deleted page.

Reply:
Is there a way to mark the parent dirty now, so that it will be included in the next checkpoint and the content cleaned up?

Reply:
That's a good point and we do that here (if we set skipp=false):
https://github.com/wiredtiger/wiredtiger/blob/8813e8bded0d6e741db25fcc8ffe2d8bbe3c703f/src/btree/bt_compact.c#L274-L277
but this doesn't seem to remove the reference to the leaf page that is in the WT_REF_DELETED state.

Reply:
Interesting - I wonder if there is something keeping that page in cache (e.g: it has other children in memory).

Will the change as proposed keep marking the page dirty, but not keep trying persistently? Or will it stop trying to get the page evicted and cleaned up at all?

Reply:
If there are no other concurrent operations and only compact is running, why would the other children pages stay in memory?

This change will stop marking the parent pages as dirty so checkpoint won't try to clean it up.

Reply:
I may misunderstand the discussion here, but my point on the internal page is dirty.

Due to a bug in checkpoint cleanup logic, we will always mark the internal page dirty when it has a fast-truncated page that is not yet obsolete to let the internal page reconcile again and again until it cleans the fast-truncated child page.

If we do not set it to dirty because of this change, it can cause disk space problems because the internal page is never reconciled.

Reply:
@kommiharibabu, are you saying we are just avoiding a real issue with those changes?
Should we focus on fixing the following:

> Due to a bug in checkpoint cleanup logic

Reply:
These are good points, and as discussed offline we should investigate why these pages are not being cleaned up by checkpoint. I've created [WT-12238](https://jira.mongodb.org/browse/WT-12238) to cover this. I will change this back to **not** skipping to ensure we mark the parent page dirty still. The rest of the changes should still apply to ensure fast truncated pages are still moved.